File: car-pooling.java
class Solution {
    //https://leetcode.com/problems/car-pooling
    public boolean carPooling(int[][] trips, int capacity) {
        Map<Integer, Integer> timestamp = new TreeMap<>();
        for (int[] trip : trips) {
            int startPassenger = timestamp.getOrDefault(trip[1], 0) + trip[0];
            timestamp.put(trip[1], startPassenger);

            int endPassenger = timestamp.getOrDefault(trip[2], 0) - trip[0];
            timestamp.put(trip[2], endPassenger);
        }
        int usedCapacity = 0;
        for (int passengerChange : timestamp.values()) {
            usedCapacity += passengerChange;
            if (usedCapacity > capacity) {
                return false;
            }
        }
        return true;
    }
}

File: car-pooling.cpp
class Solution {    
    //https://leetcode.com/problems/car-pooling
public:
    bool carPooling(vector<vector<int>>& trips, int capacity) {
        map<int, int> timestamp;
        for(auto& trip:trips) {
            int trip_1_val = (timestamp.find(trip[1]) != timestamp.end()) ? timestamp[trip[1]] : 0;
            timestamp[trip[1]] = trip_1_val + trip[0];
            int trip_2_val = (timestamp.find(trip[2]) != timestamp.end()) ? timestamp[trip[2]] : 0; 
            timestamp[trip[2]] = trip_2_val - trip[0];
        }
        
        int used_capacity = 0;
        for(auto& [key,value]:timestamp) {
            used_capacity += value;
            if( used_capacity > capacity) {
                return false;
            }
        }
        return true;   
    }
};